proposals/0366-enforce-data-complete-shred-placement.md

```diff
@@ -0,0 +1,237 @@
+   `BlockComponent` detection requires expensive searches across multiple
+   shreds and FEC sets.
+
+2. **Performance Impact**: alternative `BlockComponent` methods require either
+   complex inference algorithms or expensive replay operations.
+
+3. **Security Vulnerabilities**: Malicious leaders could intentionally
+   misplace `DATA_COMPLETE_SHRED` flags to force validators into expensive
+   search operations, creating DoS attack vectors; for certain proposed
+   `BlockComponent`s (specifically, `UpdateParent`), malicious leaders could
+   delay block repairs for following leaders.
+
+Enforcing `DATA_COMPLETE_SHRED` placement only on the last data shred in
+each FEC set allows validators to efficiently detect `BlockComponent`
+boundaries during online shred ingestion.
+
+## Forward Dependencies
+
+This proposal supports:
+
+- **[SIMD-0337]: UpdateParent Marker for Alpenglow Fast Leader Handover**
+
+    Alpenglow's fast leader handover feature depends on efficient
+    UpdateParent detection, which this SIMD enables through guaranteed FEC set
+    boundary markers.
+
+[SIMD-0337]: https://github.com/solana-foundation/solana-improvement-documents/pull/337
+
+## New Terminology
+
+- **FEC Set**: A fixed-size group of exactly 32 data shreds plus associated
+  coding shreds for erasure coding and error correction.
+
+- **FEC Set Boundary**: The transition point between consecutive FEC sets,
+  which the `DATA_COMPLETE_SHRED` flag marks on the final data shred of each
+  set.
+
+- **BlockComponent**: A control element within a block that provides metadata
+  or instructions for block processing. `BlockComponent`s align with FEC
+  set boundaries when the previous shred has `DATA_COMPLETE_SHRED` set. E.g.,
+  see https://github.com/anza-xyz/agave/pull/8100 for an implementation of
+  `BlockComponent` within Agave.
+
+## Detailed Design
+
+### FEC Set Structure
+
+FEC sets in Solana follow a fixed structure:
+
+- Each FEC set contains exactly 32 data shreds
+- Each FEC set contains exactly 32 coding shreds for error correction
+- FEC sets never allow partial fills - if insufficient data exists to fill a
+  complete set, zero-padding fills the remaining positions
+
+### `DATA_COMPLETE_SHRED` Placement Rules
+
+All validator implementations MUST enforce the following rules:
+
+1. **Placement Restriction**: The `DATA_COMPLETE_SHRED` flag MUST only appear
+   as `true` on the final data shred within an FEC set (index position 31
+   within the set, or more generally at positions where
+   `(shred_index + 1) % 32 == 0` for the data shred's position within the
+   slot).
+
+2. **Validation Requirement**: Validators MUST check `DATA_COMPLETE_SHRED`
+   placement during shred ingestion (either at turbine receipt or blockstore
+   insertion).
+
+3. **Invalid Placement Handling**: If a validator detects
+   `DATA_COMPLETE_SHRED` as `true` on any data shred other than the last
+   one in an FEC set:
+   - The validator MUST mark the entire slot as dead
```

Comment:
I think it's easier to just drop the shred, we don't need to explicitly mark the block as dead - it will fail to replay anyway.